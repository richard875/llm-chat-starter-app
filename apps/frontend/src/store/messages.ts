import { create } from "zustand";

import type { Message, Chat } from "@/types/chat";

interface MessagesState {
  messages: Message[];
  chats: Chat[];
  currentChatId: string | null;
  isLoadingChats: boolean;
  addMessage: (message: Message) => void;
  updateLastMessage: (content: string) => void;
  clearMessages: () => void;
  setCurrentChatId: (chatId: string) => void;
  startNewChat: () => void;
  fetchChats: () => Promise<void>;
  loadChatMessages: (chatId: string) => Promise<void>;
  refreshChatsAfterNewChat: () => Promise<void>;
}

export const useMessages = create<MessagesState>((set) => ({
  messages: [],
  chats: [],
  currentChatId: null,
  isLoadingChats: false,
  addMessage: (message) =>
    set((state) => ({ messages: [...state.messages, message] })),
  updateLastMessage: (content) =>
    set((state) => ({
      messages: state.messages.map((msg, idx) =>
        idx === state.messages.length - 1
          ? { ...msg, content: msg.content + content }
          : msg
      ),
    })),
  clearMessages: () => set({ messages: [] }),
  setCurrentChatId: (chatId) => set({ currentChatId: chatId }),
  startNewChat: () => {
    set({ messages: [], currentChatId: null });
  },
  fetchChats: async () => {
    set({ isLoadingChats: true });
    try {
      const response = await fetch("/api/chats");
      if (response.ok) {
        const data = await response.json();
        set({ chats: data.chats, isLoadingChats: false });
      } else {
        console.error("Failed to fetch chats");
        set({ isLoadingChats: false });
      }
    } catch (error) {
      console.error("Error fetching chats:", error);
      set({ isLoadingChats: false });
    }
  },
  loadChatMessages: async (chatId: string) => {
    try {
      const response = await fetch(`/api/messages/${chatId}`);
      if (response.ok) {
        const data = await response.json();
        set({ messages: data.messages, currentChatId: chatId });
      } else {
        console.error("Failed to fetch chat messages");
      }
    } catch (error) {
      console.error("Error fetching chat messages:", error);
    }
  },
  refreshChatsAfterNewChat: async () => {
    try {
      const response = await fetch("/api/chats");
      if (response.ok) {
        const data = await response.json();
        set({ chats: data.chats });
      }
    } catch (error) {
      console.error("Error refreshing chats:", error);
    }
  },
}));
