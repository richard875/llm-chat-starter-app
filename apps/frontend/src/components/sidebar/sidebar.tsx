import { useEffect } from "react";
import { LoaderCircle } from "lucide-react";
import { useMessages } from "@/store/messages";
import { Button } from "@/components/ui/button";

export const Sidebar = () => {
  const {
    chats,
    currentChatId,
    isLoadingMsg,
    isLoadingChats,
    startNewChat,
    fetchChats,
    loadChatMessages,
  } = useMessages();

  useEffect(() => {
    fetchChats();
  }, [fetchChats]);

  return (
    <div className="w-60 border-r shadow-sm pt-5 pb-3 p-2 gap-3 flex flex-col justify-between select-none">
      <div className="h-full overflow-auto">
        {isLoadingChats ? (
          <div className="w-full h-full flex items-center justify-center">
            <LoaderCircle className="animate-spin" />
          </div>
        ) : chats.length > 0 ? (
          chats.map((chat) => (
            <div
              className={`text-primary px-2.5 py-2 rounded-md
                ${isLoadingMsg ? "" : "cursor-pointer hover:bg-accent"}
                ${currentChatId === chat.chatId ? "bg-accent" : ""}
                `}
              key={chat.chatId}
              onClick={() => !isLoadingMsg && loadChatMessages(chat.chatId)}
            >
              <p className="text-[15px] truncate" title={chat.title}>
                {chat.title}
              </p>
            </div>
          ))
        ) : (
          <div className="w-full h-full flex items-center justify-center">
            <p className="text-[15px] text-gray-500">No Chats Yet</p>
          </div>
        )}
      </div>
      <Button
        disabled={isLoadingMsg}
        onClick={startNewChat}
        variant="outline"
        size="sm"
        className="w-full py-4.5 text-[15px] text-primary rounded-md cursor-pointer border border-primary transition bg-[#a3e636] hover:bg-[#91cc33]"
      >
        New Chat
      </Button>
    </div>
  );
};
