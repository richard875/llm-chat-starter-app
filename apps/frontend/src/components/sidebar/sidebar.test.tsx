import { describe, test, expect, vi, beforeEach } from "vitest";
import { render } from "@testing-library/react";
import { Sidebar } from "@/components/sidebar/sidebar";
import { useMessages } from "@/store/messages";
import type { Chat } from "@/types/chat";

vi.mock("@/store/messages");

describe("Sidebar", () => {
  const mockUseMessages = vi.mocked(useMessages);

  beforeEach(() => {
    vi.clearAllMocks();
    // Default mock implementation
    mockUseMessages.mockReturnValue({
      chats: [],
      currentChatId: null,
      isLoadingMsg: false,
      isLoadingChats: false,
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
      messages: [],
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
    });
  });

  test("renders loading state", () => {
    mockUseMessages.mockReturnValue({
      chats: [],
      currentChatId: null,
      isLoadingMsg: false,
      isLoadingChats: true,
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
      messages: [],
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
    });

    const { container } = render(<Sidebar />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders with chat list", () => {
    const mockChats: Chat[] = [
      {
        chatId: "chat-1",
        title: "First Conversation",
        createdAt: "2023-01-01T00:00:00.000Z",
      },
      {
        chatId: "chat-2",
        title: "Second Conversation",
        createdAt: "2023-01-02T00:00:00.000Z",
      },
    ];

    mockUseMessages.mockReturnValue({
      chats: mockChats,
      currentChatId: null,
      isLoadingMsg: false,
      isLoadingChats: false,
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
      messages: [],
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
    });

    const { container } = render(<Sidebar />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders with active chat selected", () => {
    const mockChats: Chat[] = [
      {
        chatId: "chat-1",
        title: "Active Conversation",
        createdAt: "2023-01-01T00:00:00.000Z",
      },
    ];

    mockUseMessages.mockReturnValue({
      chats: mockChats,
      currentChatId: "chat-1",
      isLoadingMsg: false,
      isLoadingChats: false,
      startNewChat: vi.fn(),
      fetchChats: vi.fn(),
      loadChatMessages: vi.fn(),
      messages: [],
      addMessage: vi.fn(),
      updateLastMessage: vi.fn(),
      setCurrentChatId: vi.fn(),
      setIsLoadingMsg: vi.fn(),
      refreshChatsAfterNewChat: vi.fn(),
    });

    const { container } = render(<Sidebar />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("renders empty state when no chats", () => {
    const { container } = render(<Sidebar />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
