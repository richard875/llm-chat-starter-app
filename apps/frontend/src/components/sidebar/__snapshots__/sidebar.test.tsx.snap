// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > renders empty state when no chats 1`] = `
<div
  class="w-60 border-r shadow-sm pt-5 pb-3 px-2 gap-3 flex flex-col justify-between select-none"
>
  <div
    class="h-full overflow-auto"
  >
    <div
      class="w-full h-full flex items-center justify-center"
    >
      <p
        class="text-[15px] text-gray-500"
      >
        No Chats Yet
      </p>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive shadow-xs hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 gap-1.5 px-3 has-[>svg]:px-2.5 w-full py-4.5 text-[15px] text-primary rounded-md cursor-pointer border border-primary transition bg-[#a3e636] hover:bg-[#91cc33]"
    data-slot="button"
  >
    New Chat
  </button>
</div>
`;

exports[`Sidebar > renders loading state 1`] = `
<div
  class="w-60 border-r shadow-sm pt-5 pb-3 px-2 gap-3 flex flex-col justify-between select-none"
>
  <div
    class="h-full overflow-auto"
  >
    <div
      class="w-full h-full flex items-center justify-center"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-loader-circle animate-spin"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 12a9 9 0 1 1-6.219-8.56"
        />
      </svg>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive shadow-xs hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 gap-1.5 px-3 has-[>svg]:px-2.5 w-full py-4.5 text-[15px] text-primary rounded-md cursor-pointer border border-primary transition bg-[#a3e636] hover:bg-[#91cc33]"
    data-slot="button"
  >
    New Chat
  </button>
</div>
`;

exports[`Sidebar > renders with active chat selected 1`] = `
<div
  class="w-60 border-r shadow-sm pt-5 pb-3 px-2 gap-3 flex flex-col justify-between select-none"
>
  <div
    class="h-full overflow-auto"
  >
    <div
      class="text-primary px-2.5 py-2 rounded-md
                cursor-pointer hover:bg-accent
                bg-accent
                "
    >
      <p
        class="text-[15px] truncate"
        title="Active Conversation"
      >
        Active Conversation
      </p>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive shadow-xs hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 gap-1.5 px-3 has-[>svg]:px-2.5 w-full py-4.5 text-[15px] text-primary rounded-md cursor-pointer border border-primary transition bg-[#a3e636] hover:bg-[#91cc33]"
    data-slot="button"
  >
    New Chat
  </button>
</div>
`;

exports[`Sidebar > renders with chat list 1`] = `
<div
  class="w-60 border-r shadow-sm pt-5 pb-3 px-2 gap-3 flex flex-col justify-between select-none"
>
  <div
    class="h-full overflow-auto"
  >
    <div
      class="text-primary px-2.5 py-2 rounded-md
                cursor-pointer hover:bg-accent
                
                "
    >
      <p
        class="text-[15px] truncate"
        title="First Conversation"
      >
        First Conversation
      </p>
    </div>
    <div
      class="text-primary px-2.5 py-2 rounded-md
                cursor-pointer hover:bg-accent
                
                "
    >
      <p
        class="text-[15px] truncate"
        title="Second Conversation"
      >
        Second Conversation
      </p>
    </div>
  </div>
  <button
    class="inline-flex items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive shadow-xs hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-8 gap-1.5 px-3 has-[>svg]:px-2.5 w-full py-4.5 text-[15px] text-primary rounded-md cursor-pointer border border-primary transition bg-[#a3e636] hover:bg-[#91cc33]"
    data-slot="button"
  >
    New Chat
  </button>
</div>
`;
